// mpesa-api-nodejs/dist/core/Request.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _axios = _interopRequireDefault(require("axios"));
var _Cryptor = _interopRequireDefault(require("./Cryptor"));
var Request = /*#__PURE__*/(0, _createClass2["default"])(function Request(api_key, public_key) {
  var _this = this;
  var ssl = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : true;
  (0, _classCallCheck2["default"])(this, Request);
  (0, _defineProperty2["default"])(this, "api_key", "");
  (0, _defineProperty2["default"])(this, "public_key", "");
  (0, _defineProperty2["default"])(this, "ssl", true);
  (0, _defineProperty2["default"])(this, "get", /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(url, params) {
      var headers, request;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            headers = {
              "Content-Type": "application/json",
              "Origin": "*",
              "Authorization": "Bearer " + _Cryptor["default"].token(_this.public_key, _this.api_key)
            };
            _context.prev = 1;
            _context.next = 4;
            return _axios["default"].get(url, {
              params: params,
              headers: headers
            });
          case 4:
            request = _context.sent;
            return _context.abrupt("return", request.data);
          case 8:
            _context.prev = 8;
            _context.t0 = _context["catch"](1);
            if (!(typeof _context.t0.response.data !== "undefined")) {
              _context.next = 12;
              break;
            }
            return _context.abrupt("return", _context.t0.response.data);
          case 12:
            throw new Error(_context.t0);
          case 13:
            throw new Error("Failed to proccess request");
          case 14:
          case "end":
            return _context.stop();
        }
      }, _callee, null, [[1, 8]]);
    }));
    return function (_x, _x2) {
      return _ref.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "post", /*#__PURE__*/function () {
    var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(url, params) {
      var length, headers, request;
      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            length = JSON.stringify(params).length || 0;
            headers = {
              "Content-Length": length,
              "Content-Type": "application/json",
              "Origin": "*",
              "Authorization": "Bearer " + _Cryptor["default"].token(_this.public_key, _this.api_key)
            };
            _context2.prev = 2;
            _context2.next = 5;
            return _axios["default"].post(url, params, {
              headers: headers
            });
          case 5:
            request = _context2.sent;
            return _context2.abrupt("return", request.data);
          case 9:
            _context2.prev = 9;
            _context2.t0 = _context2["catch"](2);
            if (!(typeof _context2.t0.response.data !== "undefined")) {
              _context2.next = 13;
              break;
            }
            return _context2.abrupt("return", _context2.t0.response.data);
          case 13:
            throw new Error(_context2.t0);
          case 14:
            throw new Error("Failed to proccess request");
          case 15:
          case "end":
            return _context2.stop();
        }
      }, _callee2, null, [[2, 9]]);
    }));
    return function (_x3, _x4) {
      return _ref2.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "put", /*#__PURE__*/function () {
    var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(url, params) {
      var length, headers, request;
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            length = JSON.stringify(params).length || 0;
            headers = {
              "Content-Length": length,
              "Content-Type": "application/json",
              "Origin": "*",
              "Authorization": "Bearer " + _Cryptor["default"].token(_this.public_key, _this.api_key)
            };
            _context3.prev = 2;
            _context3.next = 5;
            return _axios["default"].put(url, params, {
              headers: headers
            });
          case 5:
            request = _context3.sent;
            return _context3.abrupt("return", request.data);
          case 9:
            _context3.prev = 9;
            _context3.t0 = _context3["catch"](2);
            if (!(typeof _context3.t0.response.data !== "undefined")) {
              _context3.next = 13;
              break;
            }
            return _context3.abrupt("return", _context3.t0.response.data);
          case 13:
            throw new Error(_context3.t0);
          case 14:
            throw new Error("Failed to proccess request");
          case 15:
          case "end":
            return _context3.stop();
        }
      }, _callee3, null, [[2, 9]]);
    }));
    return function (_x5, _x6) {
      return _ref3.apply(this, arguments);
    };
  }());
  this.ssl = ssl;
  this.public_key = public_key;
  this.api_key = api_key;
});
;
var _default = exports["default"] = Request;
