// mpesa-api-nodejs/dist/core/Transaction.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _Request = _interopRequireDefault(require("./Request"));
var Transaction = /*#__PURE__*/(0, _createClass2["default"])(function Transaction(api_key, public_key) {
  var _this = this;
  var environment = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "development";
  var ssl = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : true;
  (0, _classCallCheck2["default"])(this, Transaction);
  (0, _defineProperty2["default"])(this, "api_key", null);
  (0, _defineProperty2["default"])(this, "public_key", null);
  (0, _defineProperty2["default"])(this, "ssl", null);
  (0, _defineProperty2["default"])(this, "is_production", false);
  (0, _defineProperty2["default"])(this, "base_url", "api.sandbox.vm.co.mz");
  (0, _defineProperty2["default"])(this, "c2b", /*#__PURE__*/function () {
    var _ref = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee$(_context) {
        while (1) switch (_context.prev = _context.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":18352/ipg/v1x/c2bPayment/singleStage/");
            params = {
              input_Amount: data.value,
              input_CustomerMSISDN: data.client_number,
              input_ServiceProviderCode: data.agent_id,
              input_TransactionReference: data.transaction_reference,
              input_ThirdPartyReference: data.third_party_reference
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context.next = 5;
            return request.post(url, params);
          case 5:
            response = _context.sent;
            return _context.abrupt("return", response);
          case 7:
          case "end":
            return _context.stop();
        }
      }, _callee);
    }));
    return function (_x) {
      return _ref.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "b2c", /*#__PURE__*/function () {
    var _ref2 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee2(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee2$(_context2) {
        while (1) switch (_context2.prev = _context2.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":18345/ipg/v1x/b2cPayment/");
            params = {
              input_Amount: data.value,
              input_CustomerMSISDN: data.client_number,
              input_ServiceProviderCode: data.agent_id,
              input_TransactionReference: data.transaction_reference,
              input_ThirdPartyReference: data.third_party_reference
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context2.next = 5;
            return request.post(url, params);
          case 5:
            response = _context2.sent;
            return _context2.abrupt("return", response);
          case 7:
          case "end":
            return _context2.stop();
        }
      }, _callee2);
    }));
    return function (_x2) {
      return _ref2.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "b2b", /*#__PURE__*/function () {
    var _ref3 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee3(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee3$(_context3) {
        while (1) switch (_context3.prev = _context3.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":18349/ipg/v1x/b2bPayment/");
            params = {
              input_PrimaryPartyCode: data.agent_id,
              input_ReceiverPartyCode: data.agent_receiver_id,
              input_Amount: data.value,
              input_TransactionReference: data.transaction_reference,
              input_ThirdPartyReference: data.third_party_reference
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context3.next = 5;
            return request.post(url, params);
          case 5:
            response = _context3.sent;
            return _context3.abrupt("return", response);
          case 7:
          case "end":
            return _context3.stop();
        }
      }, _callee3);
    }));
    return function (_x3) {
      return _ref3.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "reversal", /*#__PURE__*/function () {
    var _ref4 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee4(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee4$(_context4) {
        while (1) switch (_context4.prev = _context4.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":18354/ipg/v1x/reversal/");
            params = {
              input_TransactionID: data.transaction_id,
              input_SecurityCredential: data.security_credential,
              input_InitiatorIdentifier: data.indicator_identifier,
              input_ThirdPartyReference: data.third_party_reference,
              input_ServiceProviderCode: data.agent_id,
              input_ReversalAmount: data.value
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context4.next = 5;
            return request.put(url, params);
          case 5:
            response = _context4.sent;
            return _context4.abrupt("return", response);
          case 7:
          case "end":
            return _context4.stop();
        }
      }, _callee4);
    }));
    return function (_x4) {
      return _ref4.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "status", /*#__PURE__*/function () {
    var _ref5 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee5(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee5$(_context5) {
        while (1) switch (_context5.prev = _context5.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":18353/ipg/v1x/queryTransactionStatus/");
            params = {
              input_QueryReference: data.transaction_id,
              input_ThirdPartyReference: data.third_party_reference,
              input_ServiceProviderCode: data.agent_id
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context5.next = 5;
            return request.get(url, params);
          case 5:
            response = _context5.sent;
            return _context5.abrupt("return", response);
          case 7:
          case "end":
            return _context5.stop();
        }
      }, _callee5);
    }));
    return function (_x5) {
      return _ref5.apply(this, arguments);
    };
  }());
  (0, _defineProperty2["default"])(this, "customer_name", /*#__PURE__*/function () {
    var _ref6 = (0, _asyncToGenerator2["default"])( /*#__PURE__*/_regenerator["default"].mark(function _callee6(data) {
      var url, params, request, response;
      return _regenerator["default"].wrap(function _callee6$(_context6) {
        while (1) switch (_context6.prev = _context6.next) {
          case 0:
            url = "https://".concat(_this.base_url, ":19323/ipg/v1x/queryCustomerName/");
            params = {
              input_CustomerMSISDN: data.client_number,
              input_ThirdPartyReference: data.third_party_reference,
              input_ServiceProviderCode: data.agent_id
            };
            request = new _Request["default"](_this.api_key, _this.public_key, _this.ssl);
            _context6.next = 5;
            return request.get(url, params);
          case 5:
            response = _context6.sent;
            return _context6.abrupt("return", response);
          case 7:
          case "end":
            return _context6.stop();
        }
      }, _callee6);
    }));
    return function (_x6) {
      return _ref6.apply(this, arguments);
    };
  }());
  this.api_key = api_key;
  this.public_key = public_key;
  this.ssl = ssl;
  this.is_production = environment == "production";
  this.base_url = this.is_production ? "api.vm.co.mz" : "api.sandbox.vm.co.mz";
});
var _default = exports["default"] = Transaction;
